pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,247 @@
+	}
+
+	if err = vs.save(cvrfs); err != nil {
+		return xerrors.Errorf("error in openEuler CVRF save: %w", err)
+	}
+
+	return nil
+}
+
+func (vs VulnSrc) save(cvrfs []Cvrf) error {
+	err := vs.dbc.BatchUpdate(func(tx *bolt.Tx) error {
+		return vs.commit(tx, cvrfs)
+	})
+	if err != nil {
+		return xerrors.Errorf("error in batch update: %w", err)
+	}
+	return nil
+}
+
+func (vs VulnSrc) commit(tx *bolt.Tx, cvrfs []Cvrf) error {
+	for _, cvrf := range cvrfs {
+		affectedPkgs := getAffectedPackages(cvrf.ProductTree)
+		if len(affectedPkgs) == 0 {
+			continue
+		}
+
+		for _, affectedPkg := range affectedPkgs {
+			advisory := types.Advisory{
+				FixedVersion: affectedPkg.Package.FixedVersion,
+			}
+
+			if err := vs.dbc.PutDataSource(tx, affectedPkg.OSVer, source); err != nil {
+				return xerrors.Errorf("failed to put data source: %w", err)
+			}
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, affectedPkg.Package.Name,
+				[]string{affectedPkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", affectedPkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []AffectedPackage {
+	var pkgs []AffectedPackage
+	for _, branch := range productTree.Branches {
+		// It doesn't matter whether the packages are `x86_64` or `aarch64`,
+		// because the number and content of those two kinds of packages are the same.
+		if branch.Type != "Package Arch" || branch.Name != "aarch64" {
+			// `aarch64` is chosen because it is more typical for openEuler
+			continue
+		}
+		for _, production := range branch.Productions {
+			osVer := getOSVersion(production.CPE)
+			if osVer == "" {
+				continue
+			}
+
+			// for example
+			// "ProductID": "ignition-debuginfo-2.14.0-2"
+			// "Text": "ignition-debuginfo-2.14.0-2.oe2203sp1.aarch64.rpm"
+			productID := production.ProductID
+			if productID == "" {
+				// productID is always contained in production.Text
+				parts := strings.Split(production.Text, ".oe")
+				productID = parts[0]
+			}
```

Comment:
I found wrong `productID` - https://github.com/aquasecurity/vuln-list/blob/24612eef3897977e4671fa4d4488b60da2bc96df/openeuler/2023/openEuler-SA-2023-1410.json#L509

looks like we need to get Package from `production.ProductID`.
If package is `nil` => check `production.Text`.

Reply:
Yes, we try to get package from `production.ProductID` first. `production. Text` is used while some data are missing in SAs.

Reply:
take a look this:
```json
          {
            "ProductID": "cups-debuginfo",
            "CPE": "cpe:/a:openEuler:openEuler:22.03-LTS-SP2",
            "Text": "cups-debuginfo-2.4.0-8.oe2203sp2.x86_64.rpm"
          },
```

In this case `name` and `version` will be empty:
https://github.com/aquasecurity/trivy-db/blob/a5525f26cba264cba42a5a0137dcfd48208a05a8/pkg/vulnsrc/openeuler/openeuler.go#L216-L219

Reply:
I think this is a wrong data, let me confirm this from the community!

It seems may be `Text` should be used first?

Reply:
Using ProductID first seems more logical.
I suggest something like this:
```go
			pkg := getPackage(production.ProductID)
			if pkg == nil {
				// productID is always contained in production.Text
				parts := strings.Split(production.Text, ".oe")
				pkg = getPackage(parts[0])
			}
```

Reply:
OK